File: internal/bloblang2/eval/scope.go
package eval

// scopeMode determines how variable assignment interacts with outer scopes.
type scopeMode int

const (
	// scopeStatement: assigning to an existing outer variable modifies it.
	// New variables are block-scoped.
	scopeStatement scopeMode = iota
	// scopeExpression: assigning to an existing outer variable shadows it.
	// All variables are local to this scope.
	scopeExpression
)

// scope is a linked scope chain for variable resolution.
type scope struct {
	parent *scope
	mode   scopeMode
	vars   map[string]any
}

func newScope(parent *scope, mode scopeMode) *scope {
	return &scope{
		parent: parent,
		mode:   mode,
		vars:   make(map[string]any),
	}
}

// get looks up a variable by walking the scope chain.
func (s *scope) get(name string) (any, bool) {
	for cur := s; cur != nil; cur = cur.parent {
		if v, ok := cur.vars[name]; ok {
			return v, true
		}
	}
	return nil, false
}

// set assigns a variable, respecting the scope mode:
//   - Expression mode: always writes locally (shadow).
//   - Statement mode: if variable exists in an ancestor, update the ancestor.
//     Otherwise, create locally.
func (s *scope) set(name string, value any) {
	if s.mode == scopeStatement {
		for cur := s.parent; cur != nil; cur = cur.parent {
			if _, ok := cur.vars[name]; ok {
				cur.vars[name] = value
				return
			}
		}
	}
	s.vars[name] = value
}

// has reports whether the variable exists in this scope (not ancestors).
func (s *scope) has(name string) bool {
	_, ok := s.vars[name]
	return ok
}

File: internal/bloblang2/ts/src/scope.ts
// Scope for the Bloblang V2 interpreter.
//
// Two modes:
//   - "statement": assigning to an existing outer variable modifies it;
//     new variables are block-scoped.
//   - "expression": assigning always shadows (local); used for lambdas and maps.

import type { Value } from "./value.js";

export type ScopeMode = "statement" | "expression";

export class Scope {
  parent: Scope | null;
  mode: ScopeMode;
  vars: Map<string, Value>;

  constructor(parent: Scope | null, mode: ScopeMode) {
    this.parent = parent;
    this.mode = mode;
    this.vars = new Map();
  }

  /** Look up a variable by walking the scope chain. */
  get(name: string): Value | undefined {
    for (let cur: Scope | null = this; cur !== null; cur = cur.parent) {
      const v = cur.vars.get(name);
      if (v !== undefined) return v;
    }
    return undefined;
  }

  /**
   * Assign a variable, respecting the scope mode:
   *   - Expression mode: always writes locally (shadow).
   *   - Statement mode: if variable exists in an ancestor, update the ancestor.
   *     Otherwise, create locally.
   */
  set(name: string, value: Value): void {
    if (this.mode === "statement") {
      for (let cur = this.parent; cur !== null; cur = cur.parent) {
        if (cur.vars.has(name)) {
          cur.vars.set(name, value);
          return;
        }
      }
    }
    this.vars.set(name, value);
  }
}
